File: solutions/javascript/src/algorithm/simpledictionary/Exercise4.js
import DynamicSimpleDictionaryADT from "../../structure/dynamic/DynamicSimpleDictionaryADT.js";

function filterByValueThreshold(dictionary, threshold) {
  const result = new DynamicSimpleDictionaryADT();
  const keys = dictionary.getKeys();
  while (!keys.isEmpty()) {
    const key = keys.choose();
    keys.remove(key);
    const value = dictionary.get(key);
    if (value > threshold) {
      result.add(key, value);
    }
  }
  return result;
}


const dict = new DynamicSimpleDictionaryADT();
dict.add(1, 10);
dict.add(2, 25);
dict.add(3, 5);
dict.add(4, 30);

const filtrado = filterByValueThreshold(dict, 20);
filtrado.getKeys().mostrar();  // Suponiendo que el DynamicSetADT tenga un método mostrar()

// O para mostrar manualmente:
const clavesFiltradas = filtrado.getKeys();
while (!clavesFiltradas.isEmpty()) {
  const clave = clavesFiltradas.choose();
  clavesFiltradas.remove(clave);
  console.log(`Clave: ${clave}, Valor: ${filtrado.get(clave)}`);
}

File: solutions/python/algorithm/simpledictionary/Exercise4.py
from structure.dynamic.DynamicSimpleDictionaryADT import DynamicSimpleDictionaryADT

def filtrar_por_umbral(diccionario, umbral):
    resultado = DynamicSimpleDictionaryADT()
    for key in diccionario.get_keys():
        valor = diccionario.get(key)
        if valor > umbral:
            resultado.add(key, valor)
    return resultado

# Creamos un diccionario
dicc = DynamicSimpleDictionaryADT()
dicc.add(1, 10)
dicc.add(2, 5)
dicc.add(3, 20)
dicc.add(4, 3)

# Filtramos los valores mayores a 6
resultado = filtrar_por_umbral(dicc, 6)

# Mostramos el nuevo diccionario filtrado
resultado.mostrar()
